// react-native-mmkv/lib/commonjs/createMMKV.web.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createMMKV = void 0;
var _createTextEncoder = require("./createTextEncoder");
var _window$document;
const canUseDOM = typeof window !== 'undefined' && ((_window$document = window.document) === null || _window$document === void 0 ? void 0 : _window$document.createElement) != null;
const createMMKV = config => {
  if (config.id !== 'mmkv.default') {
    throw new Error("MMKV: 'id' is not supported on Web!");
  }
  if (config.encryptionKey != null) {
    throw new Error("MMKV: 'encryptionKey' is not supported on Web!");
  }
  if (config.path != null) {
    throw new Error("MMKV: 'path' is not supported on Web!");
  }
  const storage = () => {
    var _global, _window;
    if (!canUseDOM) {
      throw new Error('Tried to access storage on the server. Did you forget to call this in useEffect?');
    }
    const domStorage = ((_global = global) === null || _global === void 0 ? void 0 : _global.localStorage) ?? ((_window = window) === null || _window === void 0 ? void 0 : _window.localStorage) ?? localStorage;
    if (domStorage == null) {
      throw new Error(`Could not find 'localStorage' instance!`);
    }
    return domStorage;
  };
  const textEncoder = (0, _createTextEncoder.createTextEncoder)();
  return {
    clearAll: () => storage().clear(),
    delete: key => storage().removeItem(key),
    set: (key, value) => storage().setItem(key, value.toString()),
    getString: key => storage().getItem(key) ?? undefined,
    getNumber: key => {
      const value = storage().getItem(key);
      if (value == null) return undefined;
      return Number(value);
    },
    getBoolean: key => {
      const value = storage().getItem(key);
      if (value == null) return undefined;
      return value === 'true';
    },
    getBuffer: key => {
      const value = storage().getItem(key);
      if (value == null) return undefined;
      return textEncoder.encode(value);
    },
    getAllKeys: () => Object.keys(storage()),
    contains: key => storage().getItem(key) != null,
    recrypt: () => {
      throw new Error('`recrypt(..)` is not supported on Web!');
    }
  };
};
exports.createMMKV = createMMKV;
